Build command fails when symlinks in docker folder are used

My Docker image includes some files, which are symlinks pointing outside of docker folder, e.g.

```
$ ls -l include/
total 0
lrwxrwxrwx 1 dmitry dmitry 29 Mar 30 17:25 python3.5m -> /usr/local/include/python3.5m
```

Native docker client handles this correctly. Java `File.getCanonicalFile()` follows symlinks, so TAR archive sent to Docker daemon contains `usr/local/include/python3.5m/**` instead of expected `include/python3.5m/**`, and as a result I get confusing errors about non-existent files.

I guess replacing `FilePathUtil.relativize(File, File)` with the following code would work, but please check if this behavior does not break other usages.

```
    public static String relativize(File baseDir, File file) {
        try {
            return baseDir.toPath()
                          .toRealPath(LinkOption.NOFOLLOW_LINKS)
                          .relativize(file.toPath().toRealPath(LinkOption.NOFOLLOW_LINKS))
                          .toString();
        } catch (IOException e) {
            throw new DockerClientException(e.getMessage(), e);
        }
    }
```

I have another scenario with `COPY` a directory with symlinks inside. As @dvbobrov pointed, docker cli copies it just as symlinks. but using `java-docker` there's no file at all.
The normal behavior is to get a symlink.

Unfortunately, the @dvbobrov's proposal patch is not working:
```
[ERROR]   symbol:   variable LinkOption
[ERROR]   location: class com.github.dockerjava.core.util.FilePathUtil
[ERROR] /usr/src/src/main/java/com/github/dockerjava/core/util/FilePathUtil.java:[26,39] cannot find symbol
```

I'm not a java developer but I can try for a fix if someone can give me some light :)
BTW, Sorry to respond on a 2 years old bug but it's still not solved.

--- a/src/main/java/com/github/dockerjava/core/util/CompressArchiveUtil.java
+++ b/src/main/java/com/github/dockerjava/core/util/CompressArchiveUtil.java
@@ -1,7 +1,6 @@
 package com.github.dockerjava.core.util;
 
 import static com.github.dockerjava.core.util.FilePathUtil.relativize;
-import static java.nio.file.FileVisitOption.FOLLOW_LINKS;
 
 import java.io.BufferedInputStream;
 import java.io.BufferedOutputStream;
@@ -13,7 +12,6 @@ import java.io.InputStream;
 import java.io.OutputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.EnumSet;
 import java.util.zip.GZIPOutputStream;
 
 import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
@@ -30,14 +28,21 @@ public class CompressArchiveUtil {
 
     static void addFileToTar(TarArchiveOutputStream tarArchiveOutputStream, Path file, String entryName)
             throws IOException {
-        TarArchiveEntry archiveEntry = (TarArchiveEntry) tarArchiveOutputStream.createArchiveEntry(file.toFile(), entryName);
-        if (file.toFile().canExecute()) {
-            archiveEntry.setMode(archiveEntry.getMode() | 0755);
-        }
-        tarArchiveOutputStream.putArchiveEntry(archiveEntry);
-        if (file.toFile().isFile()) {
-            try (InputStream input = new BufferedInputStream(Files.newInputStream(file))) {
-                ByteStreams.copy(input, tarArchiveOutputStream);
+        if (Files.isSymbolicLink(file)) {
+            TarArchiveEntry tarArchiveEntry = new TarArchiveEntry(entryName, TarArchiveEntry.LF_SYMLINK);
+            tarArchiveEntry.setLinkName(Files.readSymbolicLink(file).toString());
+            tarArchiveOutputStream.putArchiveEntry(tarArchiveEntry);
+        } else {
+            TarArchiveEntry tarArchiveEntry = (TarArchiveEntry) tarArchiveOutputStream.createArchiveEntry(file.toFile(),
+                    entryName);
+            if (file.toFile().canExecute()) {
+                tarArchiveEntry.setMode(tarArchiveEntry.getMode() | 0755);
+            }
+            tarArchiveOutputStream.putArchiveEntry(tarArchiveEntry);
+            if (file.toFile().isFile()) {
+                try (InputStream input = new BufferedInputStream(Files.newInputStream(file))) {
+                    ByteStreams.copy(input, tarArchiveOutputStream);
+                }
             }
         }
         tarArchiveOutputStream.closeArchiveEntry();
@@ -81,8 +86,7 @@ public class CompressArchiveUtil {
                     // In order to have the dossier as the root entry
                     sourcePath = inputPath.getParent();
                 }
-                Files.walkFileTree(inputPath, EnumSet.of(FOLLOW_LINKS), Integer.MAX_VALUE,
-                        new TarDirWalker(sourcePath, tarArchiveOutputStream));
+                Files.walkFileTree(inputPath, new TarDirWalker(sourcePath, tarArchiveOutputStream));
             }
             tarArchiveOutputStream.flush();
         }
@@ -96,7 +100,10 @@ public class CompressArchiveUtil {
                 new FileOutputStream(tarFile))))) {
             tos.setLongFileMode(TarArchiveOutputStream.LONGFILE_GNU);
             for (File file : files) {
-                addFileToTar(tos, file.toPath(), relativize(base, file));
+                // relativize with method using Path otherwise method with File resolves the symlinks
+                // and this is not want we want. If the file is a symlink, the relativized path should
+                // keep the symlink name and not the target it points to.
+                addFileToTar(tos, file.toPath(), relativize(base.toPath(), file.toPath()));
             }
         }
 

--- a/src/main/java/com/github/dockerjava/core/util/CompressArchiveUtil.java
+++ b/src/main/java/com/github/dockerjava/core/util/CompressArchiveUtil.java
@@ -28,14 +28,19 @@ public class CompressArchiveUtil {
         // utility class
     }
 
-    static void putTarEntry(TarArchiveOutputStream tarOutputStream, TarArchiveEntry tarEntry, Path file)
+    static void addFileToTar(TarArchiveOutputStream tarArchiveOutputStream, Path file, String entryName)
             throws IOException {
-        tarEntry.setSize(Files.size(file));
-        tarOutputStream.putArchiveEntry(tarEntry);
-        try (InputStream input = new BufferedInputStream(Files.newInputStream(file))) {
-            ByteStreams.copy(input, tarOutputStream);
-            tarOutputStream.closeArchiveEntry();
+        TarArchiveEntry archiveEntry = (TarArchiveEntry) tarArchiveOutputStream.createArchiveEntry(file.toFile(), entryName);
+        if (file.toFile().canExecute()) {
+            archiveEntry.setMode(archiveEntry.getMode() | 0755);
         }
+        tarArchiveOutputStream.putArchiveEntry(archiveEntry);
+        if (file.toFile().isFile()) {
+            try (InputStream input = new BufferedInputStream(Files.newInputStream(file))) {
+                ByteStreams.copy(input, tarArchiveOutputStream);
+            }
+        }
+        tarArchiveOutputStream.closeArchiveEntry();
     }
 
     private static TarArchiveOutputStream buildTarStream(Path outputPath, boolean gZipped) throws IOException {
@@ -69,11 +74,7 @@ public class CompressArchiveUtil {
 
         try (TarArchiveOutputStream tarArchiveOutputStream = buildTarStream(outputPath, gZipped)) {
             if (!Files.isDirectory(inputPath)) {
-                TarArchiveEntry tarEntry = new TarArchiveEntry(inputPath.getFileName().toString());
-                if (inputPath.toFile().canExecute()) {
-                    tarEntry.setMode(tarEntry.getMode() | 0755);
-                }
-                putTarEntry(tarArchiveOutputStream, tarEntry, inputPath);
+                addFileToTar(tarArchiveOutputStream, inputPath, inputPath.getFileName().toString());
             } else {
                 Path sourcePath = inputPath;
                 if (!childrenOnly) {
@@ -95,19 +96,7 @@ public class CompressArchiveUtil {
                 new FileOutputStream(tarFile))))) {
             tos.setLongFileMode(TarArchiveOutputStream.LONGFILE_GNU);
             for (File file : files) {
-                TarArchiveEntry tarEntry = new TarArchiveEntry(file);
-                tarEntry.setName(relativize(base, file));
-
-                if (!file.isDirectory() && file.canExecute()) {
-                    tarEntry.setMode(tarEntry.getMode() | 0755);
-                }
-
-                tos.putArchiveEntry(tarEntry);
-
-                if (!file.isDirectory()) {
-                    FileUtils.copyFile(file, tos);
-                }
-                tos.closeArchiveEntry();
+                addFileToTar(tos, file.toPath(), relativize(base, file));
             }
         }
 

--- a/src/main/java/com/github/dockerjava/core/util/TarDirWalker.java
+++ b/src/main/java/com/github/dockerjava/core/util/TarDirWalker.java
@@ -33,9 +33,6 @@ public class TarDirWalker extends SimpleFileVisitor<Path> {
 
     @Override
     public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
-        if (attrs.isSymbolicLink()) { // symbolic link to folder
-            return FileVisitResult.CONTINUE;
-        }
         CompressArchiveUtil.addFileToTar(tarArchiveOutputStream, file, FilePathUtil.relativize(basePath, file));
         return FileVisitResult.CONTINUE;
     }

--- a/src/main/java/com/github/dockerjava/core/util/TarDirWalker.java
+++ b/src/main/java/com/github/dockerjava/core/util/TarDirWalker.java
@@ -36,11 +36,7 @@ public class TarDirWalker extends SimpleFileVisitor<Path> {
         if (attrs.isSymbolicLink()) { // symbolic link to folder
             return FileVisitResult.CONTINUE;
         }
-        TarArchiveEntry tarEntry = new TarArchiveEntry(FilePathUtil.relativize(basePath, file));
-        if (file.toFile().canExecute()) {
-            tarEntry.setMode(tarEntry.getMode() | 0755);
-        }
-        CompressArchiveUtil.putTarEntry(tarArchiveOutputStream, tarEntry, file);
+        CompressArchiveUtil.addFileToTar(tarArchiveOutputStream, file, FilePathUtil.relativize(basePath, file));
         return FileVisitResult.CONTINUE;
     }
 

--- a/src/test/java/com/github/dockerjava/core/util/CompressArchiveUtilTest.java
+++ b/src/test/java/com/github/dockerjava/core/util/CompressArchiveUtilTest.java
@@ -2,7 +2,6 @@ package com.github.dockerjava.core.util;
 
 import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
 import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
-import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
@@ -69,7 +68,6 @@ public class CompressArchiveUtilTest {
     }
 
     @Test
-    @Ignore("Symlink creation is broken so test do not pass")
     public void tarWithSymbolicLinkFileAsInput() throws IOException {
         Path archiveSourceFile = tempFolder.getRoot().toPath().resolve("symlinkFile");
         Path linkTargetFile = tempFolder.newFile("link-target").toPath();
@@ -140,7 +138,6 @@ public class CompressArchiveUtilTest {
     }
 
     @Test
-    @Ignore("Symlink creation is broken so test do not pass")
     public void tarWithfolderAsInputAndNestedSymbolicLinkFile() throws Exception {
         Path archiveSourceDir = tempFolder.newFolder("archive-source").toPath();
         Path linkTargetFile = tempFolder.newFile("link-target").toPath();
@@ -162,7 +159,6 @@ public class CompressArchiveUtilTest {
     }
 
     @Test
-    @Ignore("Symlink creation is broken so test do not pass")
     public void tarWithfolderAsInputAndNestedSymbolicLinkDir() throws Exception {
         Path archiveSourceDir = tempFolder.newFolder("archive-source").toPath();
         Path linkTargetDir = tempFolder.newFolder("link-target").toPath();
@@ -207,7 +203,6 @@ public class CompressArchiveUtilTest {
     }
 
     @Test
-    @Ignore("Symlink creation is broken so test do not pass")
     public void archiveTARFilesWithSymbolicLinkFile() throws Exception {
         Path linkTargetFile = tempFolder.newFile("link-target").toPath();
         Path symlinkFile = tempFolder.getRoot().toPath().resolve("symlinkFile");
@@ -219,7 +214,6 @@ public class CompressArchiveUtilTest {
     }
 
     @Test
-    @Ignore("Symlink creation is broken so test do not pass")
     public void archiveTARFilesWithSymbolicLinkDir() throws Exception {
         Path linkTargetDir = tempFolder.newFolder("link-target").toPath();
         Path symlinkFile = tempFolder.getRoot().toPath().resolve("symlinkFile");

